File: apps/agent-runtime/src/platform/core/contracts.ts
import { z } from 'zod';

export const PLATFORM_CONTRACT_VERSION = 1 as const;

export const PlatformIdSchema = z.enum(['kook', 'telegram', 'wechat', 'whatsapp']);
export type PlatformId = z.infer<typeof PlatformIdSchema>;

export const ChatTypeSchema = z.enum(['private', 'group']);
export type ChatType = z.infer<typeof ChatTypeSchema>;

export const PlatformIdentitySchema = z.object({
  platform: PlatformIdSchema,
  platformUserId: z.string().min(1),
  internalUserId: z.string().min(1).nullable(),
  displayName: z.string().nullable(),
});
export type PlatformIdentity = z.infer<typeof PlatformIdentitySchema>;

export const NormalizedChatSchema = z.object({
  type: ChatTypeSchema,
  id: z.string().min(1),
  name: z.string().nullable(),
});
export type NormalizedChat = z.infer<typeof NormalizedChatSchema>;

export const NormalizedMentionSchema = z.object({
  platformUserId: z.string().min(1),
  displayName: z.string().nullable(),
});
export type NormalizedMention = z.infer<typeof NormalizedMentionSchema>;

export const NormalizedAttachmentSchema = z.object({
  type: z.string().min(1),
  url: z.string().nullable(),
  name: z.string().nullable(),
  mimeType: z.string().nullable(),
});
export type NormalizedAttachment = z.infer<typeof NormalizedAttachmentSchema>;

export const NormalizedBotMessageSchema = z.object({
  version: z.literal(PLATFORM_CONTRACT_VERSION),
  platform: PlatformIdSchema,
  botId: z.string().min(1),
  user: PlatformIdentitySchema,
  chat: NormalizedChatSchema,
  message: z.object({
    id: z.string().min(1),
    text: z.string(),
    replyToMessageId: z.string().nullable(),
  }),
  mentions: z.array(NormalizedMentionSchema),
  attachments: z.array(NormalizedAttachmentSchema),
  timestamp: z.string().min(1),
  callback: z.object({
    id: z.string().min(1).nullable(),
    data: z.string().min(1),
  }).optional(),
  raw: z.unknown().optional(),
});
export type NormalizedBotMessage = z.infer<typeof NormalizedBotMessageSchema>;

export const PlatformCapabilitiesSchema = z.object({
  supportsMarkdown: z.boolean(),
  supportsCodeBlock: z.boolean(),
  supportsReply: z.boolean(),
  supportsImages: z.boolean(),
  supportsFiles: z.boolean(),
  supportsStreaming: z.boolean(),
  supportsButtons: z.boolean(),
  maxMessageLength: z.number().int().positive(),
});
export type PlatformCapabilities = z.infer<typeof PlatformCapabilitiesSchema>;

export const PresentationSectionSchema = z.object({
  type: z.string().min(1),
  title: z.string().optional(),
  text: z.string().optional(),
  lines: z.array(z.string()).optional(),
  data: z.record(z.string(), z.unknown()).optional(),
});
export type PresentationSection = z.infer<typeof PresentationSectionSchema>;

export const PresentationModelSchema = z.object({
  version: z.literal(1),
  type: z.string().min(1),
  title: z.string().optional(),
  subtitle: z.string().optional(),
  sections: z.array(PresentationSectionSchema),
  fallbackText: z.string(),
  metadata: z.record(z.string(), z.unknown()).default({}),
});
export type PresentationModel = z.infer<typeof PresentationModelSchema>;

export const BotResponseMessageSchema = z.object({
  type: z.enum(['text', 'image', 'file']),
  text: z.string().optional(),
  url: z.string().optional(),
  name: z.string().optional(),
  buttons: z.array(z.object({
    text: z.string().min(1),
    callbackData: z.string().min(1),
  })).optional(),
});
export type BotResponseMessage = z.infer<typeof BotResponseMessageSchema>;

export const BotResponseSchema = z.object({
  messages: z.array(BotResponseMessageSchema),
  replyTo: z.string().nullable(),
  metadata: z.record(z.string(), z.unknown()).default({}),
});
export type BotResponse = z.infer<typeof BotResponseSchema>;

export interface PlatformAdapter<TRawEvent = unknown> {
  readonly platform: PlatformId;
  normalize(event: TRawEvent): NormalizedBotMessage;
}

export interface PlatformRenderer {
  readonly platform: PlatformId;
  readonly capabilities: PlatformCapabilities;
  render(presentation: PresentationModel, message?: NormalizedBotMessage): BotResponse;
}

export interface PlatformSender<TTarget = unknown> {
  send(target: TTarget, response: BotResponse): Promise<void>;
}

const PLATFORM_ALIASES: Record<string, PlatformId> = {
  kook: 'kook',
  'kook-bot': 'kook',
  KOOK: 'kook',
  telegram: 'telegram',
  'telegram-bot': 'telegram',
  tg: 'telegram',
  wechat: 'wechat',
  wx: 'wechat',
  whatsapp: 'whatsapp',
  'whatsapp-cloud': 'whatsapp',
  'whatsapp-business': 'whatsapp',
  wa: 'whatsapp',
};

export function normalizePlatform(value: unknown): PlatformId {
  const normalized = String(value ?? '').trim();
  const platform = PLATFORM_ALIASES[normalized] ?? PLATFORM_ALIASES[normalized.toLowerCase()];
  if (!platform) throw new Error(`unsupported platform: ${normalized || 'empty'}`);
  return platform;
}

export function normalizeChatType(value: unknown): ChatType {
  const normalized = String(value ?? '').trim().toLowerCase();
  if (['private', 'person', 'direct', 'dm', 'user'].includes(normalized)) return 'private';
  return 'group';
}

File: integrations/langbot/plugins/pubg-stats-v3/components/platform/contracts.py
from __future__ import annotations

from typing import Any, TypedDict

NORMALIZED_MESSAGE_VERSION = 1
PLATFORMS = ('kook', 'telegram', 'wechat', 'whatsapp')
CHAT_TYPES = ('private', 'group')


class NormalizedUser(TypedDict):
    platform: str
    platformUserId: str
    internalUserId: str | None
    displayName: str | None


class NormalizedChat(TypedDict):
    type: str
    id: str
    name: str | None


class NormalizedMessageInfo(TypedDict):
    id: str
    text: str
    replyToMessageId: str | None


class NormalizedCallback(TypedDict):
    id: str | None
    data: str


class NormalizedBotMessage(TypedDict, total=False):
    version: int
    platform: str
    botId: str
    user: NormalizedUser
    chat: NormalizedChat
    message: NormalizedMessageInfo
    mentions: list[dict[str, Any]]
    attachments: list[dict[str, Any]]
    timestamp: str
    callback: NormalizedCallback


def _value(value: Any, default: str = '') -> str:
    raw = getattr(value, 'value', value)
    if raw is None or raw == '':
        return default
    return str(raw)


def normalize_platform(value: Any) -> str:
    normalized = _value(value).strip().lower()
    aliases = {
        'kook': 'kook',
        'kook-bot': 'kook',
        'telegram': 'telegram',
        'telegram-bot': 'telegram',
        'tg': 'telegram',
        'wechat': 'wechat',
        'wx': 'wechat',
        'whatsapp': 'whatsapp',
        'whatsapp-cloud': 'whatsapp',
        'whatsapp-business': 'whatsapp',
        'wa': 'whatsapp',
    }
    platform = aliases.get(normalized)
    if platform is None:
        raise ValueError(f'unsupported platform: {normalized or "empty"}')
    return platform


def normalize_chat_type(value: Any) -> str:
    normalized = _value(value).strip().lower()
    return 'private' if normalized in {'private', 'person', 'direct', 'dm', 'user'} else 'group'


def _mapping_value(value: Any, key: str, default: Any = None) -> Any:
    if isinstance(value, dict):
        return value.get(key, default)
    return getattr(value, key, default)


def safe_text(value: Any) -> str:
    return _value(value).strip()


def safe_timestamp(value: Any) -> str:
    if value is None or value == '':
        from datetime import datetime, timezone

        return datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z')
    numeric = None
    try:
        numeric = float(value)
    except (TypeError, ValueError):
        pass
    if numeric is not None:
        from datetime import datetime, timezone

        seconds = numeric / 1000 if numeric > 10_000_000_000 else numeric
        return datetime.fromtimestamp(seconds, timezone.utc).isoformat().replace('+00:00', 'Z')
    return str(value)


def build_normalized_message(
    *,
    platform: Any,
    bot_id: Any,
    platform_user_id: Any,
    internal_user_id: Any = None,
    display_name: Any = None,
    chat_type: Any,
    chat_id: Any,
    chat_name: Any = None,
    message_id: Any,
    text: Any,
    reply_to_message_id: Any = None,
    timestamp: Any = None,
    callback_id: Any = None,
    callback_data: Any = None,
) -> NormalizedBotMessage:
    canonical_platform = normalize_platform(platform)
    message: NormalizedBotMessage = {
        'version': NORMALIZED_MESSAGE_VERSION,
        'platform': canonical_platform,
        'botId': safe_text(bot_id) or f'{canonical_platform}-bot',
        'user': {
            'platform': canonical_platform,
            'platformUserId': safe_text(platform_user_id) or 'unknown',
            'internalUserId': safe_text(internal_user_id) or None,
            'displayName': safe_text(display_name) or None,
        },
        'chat': {
            'type': normalize_chat_type(chat_type),
            'id': safe_text(chat_id) or 'unknown',
            'name': safe_text(chat_name) or None,
        },
        'message': {
            'id': safe_text(message_id) or 'unknown-message',
            'text': safe_text(text),
            'replyToMessageId': safe_text(reply_to_message_id) or None,
        },
        'mentions': [],
        'attachments': [],
        'timestamp': safe_timestamp(timestamp),
    }
    if callback_data:
        message['callback'] = {'id': safe_text(callback_id) or None, 'data': safe_text(callback_data)}
    return message


def mapping_value(value: Any, key: str, default: Any = None) -> Any:
    return _mapping_value(value, key, default)
